### PR TITLE
get language and version from detected services

### PR DESCRIPTION
**Pull request by DanielMcSheehy**
how to test: 
- put a new dir in backeffx/services (backeffx/services/ghost)
- in the new dir put a main.go file (can be empty)
- run `go run effx.go sync -d ../backeffx  -k $EFFX_API_KEY`

```
go run effx.go sync -d ../backeffx  -k $EFFX_API_KEY
detected service name ghost
detected service tags map[go:1.13 language:go]
2021/03/01 12:47:08 Successfully detected  1  service(s)
2021/03/01 12:47:08 Syncing ../backeffx/effx.yaml
```

### CODE REVIEW

File: data/data.go [hunk 1]
```diff
@@ -53,9 +53,11 @@ func setMetadata(config *effx_api.ConfigurationFile, m *metadata.Result) *effx_a
 		}
 
 		if m.Language != "" && m.Version != "" {
+			language := strings.ToLower(m.Language)
+			version := strings.ToLower(m.Version)
 			safelySetField(config.Annotations, "effx.io/inferred-tags", fmt.Sprintf("language,%s", m.Language))
-			safelySetField(config.Tags, "language", strings.ToLower(m.Language))
-			safelySetField(config.Tags, m.Language, strings.ToLower(m.Version))
```
Comment (DanielMcSheehy) [inline data/data.go:58]: this was a bug I found

File: data/data.go [hunk 2]
```diff
@@ -53,9 +53,11 @@ func setMetadata(config *effx_api.ConfigurationFile, m *metadata.Result) *effx_a
 		}
 
 		if m.Language != "" && m.Version != "" {
+			language := strings.ToLower(m.Language)
+			version := strings.ToLower(m.Version)
 			safelySetField(config.Annotations, "effx.io/inferred-tags", fmt.Sprintf("language,%s", m.Language))
```
Comment (weeyum) [inline data/data.go:58]: i think the logic here could be better a little bit. perhaps we should allow inferring one of these tags by itself, instead of requiring both. Also, we might consider not setting this field for detected services

Comment (DanielMcSheehy) [reply data/data.go]: @weeyum So Joey mentioned a while back that having a language tag without a version is pointless. How do you feel about this? 

I just added a change that allows the language field to be set without a version. Also this function is not called for detected services.

Comment (weeyum) [reply data/data.go]: right, but the idea is that the logic would only be skipped if someone already defined it. so it would still have it, but i'm saying that we should allow a user to define 1 of the 2 tags

Comment (weeyum) [reply data/data.go]: for example

customer:
language: go

inferred
go:1.13.5

I would expect annotations to have inferred-tags:go

File: data/data.go [hunk 3]
```diff
@@ -45,17 +45,24 @@ func safelySetField(configMap *map[string]string, field, value string) {
...
+
 		if config.Annotations == nil {
 			config.Annotations = &map[string]string{}
 		}
 		if config.Tags == nil {
 			config.Tags = &map[string]string{}
 		}
 
-		if m.Language != "" && m.Version != "" {
+		if language != "" {
```
Comment (weeyum) [inline data/data.go:58]: we should do this for version too

Comment (weeyum) [reply data/data.go:58]: but i dont think the logic would work as you have set it up

File: data/data.go (hunk 2 above)
Comment (weeyum) [reply data/data.go]: @DanielMcSheehy see this comment ^

Comment (weeyum) [reply data/data.go]: but also we should make it work for the inverse, where they define `go:...` and we infer `language:go`

Comment (euriostigue) [reply data/data.go]: I'm fixing this in a different PR

File: internal/parser/process.go [hunk 4]
```diff
@@ -61,16 +61,10 @@ func ProcessDirectory(directory string) []data.EffxYaml {
 	return yamls
 }
 
-func DetectServicesFromEffxYamls(files []data.EffxYaml, apiKeyString, sourceName string) error {
-	filePaths := []string{}
+func DetectServicesFromEffxYamls(filePaths []string, apiKeyString, sourceName string) error {
```
Comment (euriostigue) [inline internal/parser/process.go:64]: Why did you need to change the function signature here? I think it was cleaner before

File: internal/discover/discover.go [hunk 5]
```diff
@@ -86,21 +89,32 @@ func DetectServices(effxFileLocations []string) []string {
...
+					if err != nil {
+						log.Printf("Could not predict version %+v\n", err)
+					}
+
+					if result != nil && result.Language != "" && result.Version != "" {
+						payload.Tags = &map[string]string{}
+						(*payload.Tags)["language"] = strings.ToLower(result.Language)
+						(*payload.Tags)[strings.ToLower(result.Language)] = strings.ToLower(result.Version)
+					}
+				}
```
Comment (euriostigue) [inline internal/discover/discover.go:107]: I think you would want to break this up into two sections.
1) If you detected a language
2) If you detected both language and version

As is, you would omit language suggestions if version detection failed.

**Review by euriostigue — CHANGES_REQUESTED** [review]

File: internal/parser/process.go (hunk 4 above)
Comment (DanielMcSheehy) [reply internal/parser/process.go:64]: this makes it easier for vcs connect to import and use this file. I agree it was cleaner before but vcs-connect doesn't use the data.Effxyaml datatype

File: internal/discover/discover.go (hunk 5 above)
Comment (DanielMcSheehy) [reply internal/discover/discover.go]: Good call, now if version cannot be detect but language can, the language tag will be used.

Comment (DanielMcSheehy) [reply internal/discover/discover.go]: fyi, version cannot be found if a language cannot be detected.

File: internal/parser/process.go (hunk 4 above)
Comment (euriostigue) [reply internal/parser/process.go:64]: Can `vcs-connect` use `discover.DetectServices` directly?

Comment (euriostigue) [reply internal/parser/process.go:64]: I haven't read through `vcs-connect` but I'd think if you want to use utilities in `parser`, you should start with `ProcessDirectory` so that you start with appropriate data types.

Comment (DanielMcSheehy) [reply internal/parser/process.go]: did not think about that, you're right. Cool, reverted it back

**Review by euriostigue — APPROVED** [review]
